text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,252 @@
+- take 1 bytes from `reserved1` for a new `version: u8` field.
+- take 2 bytes from `reserved1` for a new `core_index: u16` field.
+- take 4 bytes from `reserved1` for a new `session_index: u32` field.
+- the remaining `reserved1` and `reserved2` fields are zeroed
+
+Thew new primitive will look like this:
+
+```rust
+pub struct CandidateDescriptorV2<H = Hash> {
+	/// The ID of the para this is a candidate for.
+	para_id: ParaId,
+	/// The hash of the relay-chain block this is executed in the context of.
+	relay_parent: H,
+	/// Version field. The raw value here is not exposed, instead it is used
+	/// to determine the `CandidateDescriptorVersion`, see `fn version()`
+	version: InternalVersion,
+	/// The core index where the candidate is backed.
+	core_index: u16,
+	/// The session index of the candidate relay parent.
+	session_index: SessionIndex,
+	/// Reserved bytes.
+	reserved25b: [u8; 25],
+	/// The blake2-256 hash of the persisted validation data. This is extra data derived from
+	/// relay-chain state which may vary based on bitfields included before the candidate.
+	/// Thus it cannot be derived entirely from the relay-parent.
+	persisted_validation_data_hash: Hash,
+	/// The blake2-256 hash of the PoV.
+	pov_hash: Hash,
+	/// The root of a block's erasure encoding Merkle tree.
+	erasure_root: Hash,
+	/// Reserved bytes.
+	reserved64b: [u8; 64],
+	/// Hash of the para header that is being generated by this candidate.
+	para_head: Hash,
+	/// The blake2-256 hash of the validation code bytes.
+	validation_code_hash: ValidationCodeHash,
+}
+```
+
+In future format versions, parts of the `reserved1` and `reserved2` bytes can be used to include additional information in the descriptor.
+
+#### Candidate descriptor API
+
+We want to decouple the actual representation of the `CandidateDescriptor` from the higher level code. This should make it easier to implement future format versions of this primitive. To hide the logic of versioning the descriptor fields will be private and getter methods are provided for all the fields. 
+
+```rust
+impl<H> CandidateDescriptorV2<H> {
+
+	/// Returns the collator id in the descriptor. Returns `None` if descriptor is at verision 2.
+	pub fn collator(&self) -> Option<CollatorId>;
+
+	/// Returns the collator signature in the descriptor. Returns `None` if descriptor is at verision 2.
+	pub fn signature(&self) -> Option<CollatorSignature>;
+
+	/// Returns the core index of the descriptor. Returns `None` if the descriptor is at version 1.
+	pub fn core_index(&self) -> Option<CoreIndex>;
+
+	/// Returns the session index of the descriptor. Returns `None` if the descriptor is at version 1.
+	pub fn session_index(&self) -> Option<SessionIndex>;
+
+	/// ...
+}
+
+```
+
+A manual decode `Decode`  implementation is required to account for version detection and constructing the appropriate variant.
+
+### Parachain block validation
```

Comment:
should we have a similar section about how collators are going to use it?

Reply:
Not in this section, I will add something more explicit in the `Ergonomics` section.

Reply:
I've added an example along with the `UMPSignal::SelectCore` definition